File: src/metrics/duplication.zig
const std = @import("std");
const tree_sitter = @import("../parser/tree_sitter.zig");
const halstead = @import("halstead.zig");

/// A single normalized token extracted from an AST leaf node.
/// The token identity for hashing is `kind` (normalized type string).
/// For identifier nodes, kind is replaced with sentinel "V" for Type 2 clone detection.
pub const Token = struct {
    /// Normalized token type or sentinel "V" for identifiers.
    kind: []const u8,
    /// Byte offset in source (for line-number lookup).
    start_byte: u32,
    /// 1-indexed line number (pre-computed from TSNode).
    start_line: u32,
};

/// A window of consecutive tokens identified by a rolling hash.
pub const TokenWindow = struct {
    /// Index into the file_paths array.
    file_index: u32,
    /// Start token index in the file's token sequence.
    start_token: u32,
    /// Exclusive end token index.
    end_token: u32,
    /// 1-indexed line number of first token.
    start_line: u32,
    /// 1-indexed line number of last token.
    end_line: u32,
};

/// A detected clone group: two or more locations with the same normalized token sequence.
pub const CloneGroup = struct {
    /// Number of tokens in the clone window.
    token_count: u32,
    /// All instances of this clone (2+).
    locations: []const CloneLocation,
};

/// A single instance of a clone at a specific file location.
pub const CloneLocation = struct {
    /// Relative file path.
    file_path: []const u8,
    /// 1-indexed start line.
    start_line: u32,
    /// 1-indexed end line.
    end_line: u32,
};

/// Per-file duplication summary.
pub const FileDuplicationResult = struct {
    /// Relative file path.
    path: []const u8,
    /// Total token count in this file.
    total_tokens: u32,
    /// Non-overlapping count of tokens participating in any clone.
    cloned_tokens: u32,
    /// cloned_tokens / total_tokens * 100
    duplication_pct: f64,
    /// True when duplication_pct >= file_warning_pct.
    warning: bool,
    /// True when duplication_pct >= file_error_pct.
    @"error": bool,
};

/// Project-wide duplication result.
pub const DuplicationResult = struct {
    /// All detected clone groups (2+ locations each).
    clone_groups: []const CloneGroup,
    /// Per-file duplication summaries.
    file_results: []const FileDuplicationResult,
    /// Sum of non-overlapping cloned tokens across all files.
    total_cloned_tokens: u32,
    /// Sum of all tokens across all files.
    total_tokens: u32,
    /// total_cloned_tokens / total_tokens * 100
    project_duplication_pct: f64,
    /// True when project_duplication_pct >= project_warning_pct.
    project_warning: bool,
    /// True when project_duplication_pct >= project_error_pct.
    project_error: bool,
};

/// Configuration for duplication detection.
pub const DuplicationConfig = struct {
    /// Minimum number of tokens for a clone window (default 25).
    min_window: u32,
    /// File-level warning threshold percentage.
    file_warning_pct: f64,
    /// File-level error threshold percentage.
    file_error_pct: f64,
    /// Project-level warning threshold percentage.
    project_warning_pct: f64,
    /// Project-level error threshold percentage.
    project_error_pct: f64,

    /// Returns default configuration with industry-standard thresholds.
    pub fn default() DuplicationConfig {
        return DuplicationConfig{
            .min_window = 25,
            .file_warning_pct = 15.0,
            .file_error_pct = 25.0,
            .project_warning_pct = 5.0,
            .project_error_pct = 10.0,
        };
    }
};

/// Input to detectDuplication: a file path and its token sequence.
pub const FileTokens = struct {
    /// Relative file path.
    path: []const u8,
    /// Normalized token sequence for this file.
    tokens: []const Token,
};

/// Tokenize a tree-sitter AST into a normalized token sequence.
/// Strips comments, whitespace, and TypeScript type annotations.
/// Normalizes identifier nodes to sentinel "V" for Type 2 clone detection.
pub fn tokenizeTree(
    allocator: std.mem.Allocator,
    root: tree_sitter.Node,
    source: []const u8,
) ![]Token {
    _ = source;
    var tokens = std.ArrayList(Token).empty;
    errdefer tokens.deinit(allocator);
    try tokenizeNode(root, &tokens, allocator);
    return try tokens.toOwnedSlice(allocator);
}

/// Detect code clones across multiple files using Rabin-Karp rolling hash.
/// Returns clone groups with per-file and project-level duplication statistics.
pub fn detectDuplication(
    allocator: std.mem.Allocator,
    file_tokens: []const FileTokens,
    config: DuplicationConfig,
) !DuplicationResult {
    _ = file_tokens;
    _ = config;
    // Stub: return empty result (tests will fail in RED phase)
    return DuplicationResult{
        .clone_groups = try allocator.alloc(CloneGroup, 0),
        .file_results = try allocator.alloc(FileDuplicationResult, 0),
        .total_cloned_tokens = 0,
        .total_tokens = 0,
        .project_duplication_pct = 0.0,
        .project_warning = false,
        .project_error = false,
    };
}

// --- Internal helpers (stubs for RED phase) ---

fn tokenizeNode(
    node: tree_sitter.Node,
    tokens: *std.ArrayList(Token),
    allocator: std.mem.Allocator,
) !void {
    _ = node;
    _ = tokens;
    _ = allocator;
    // Stub: does nothing — tests will fail because token count == 0
}

// TESTS

test "tokenizeTree: produces tokens from simple function" {
    const parser = try tree_sitter.Parser.init();
    defer parser.deinit();
    try parser.setLanguage(.typescript);

    const source = "function add(a: number, b: number): number { return a + b; }";
    const tree = try parser.parseString(source);
    defer tree.deinit();

    const root = tree.rootNode();
    const tokens = try tokenizeTree(std.testing.allocator, root, source);
    defer std.testing.allocator.free(tokens);

    // Should produce more than 0 tokens from a real function
    try std.testing.expect(tokens.len > 0);
}

test "tokenizeTree: skips comments" {
    const parser = try tree_sitter.Parser.init();
    defer parser.deinit();
    try parser.setLanguage(.typescript);

    const source =
        \\// This is a comment
        \\function greet(name: string): string {
        \\  /* block comment */
        \\  return name;
        \\}
    ;
    const tree = try parser.parseString(source);
    defer tree.deinit();

    const root = tree.rootNode();
    const tokens = try tokenizeTree(std.testing.allocator, root, source);
    defer std.testing.allocator.free(tokens);

    // No token should have kind "comment", "line_comment", or "block_comment"
    for (tokens) |tok| {
        try std.testing.expect(!std.mem.eql(u8, tok.kind, "comment"));
        try std.testing.expect(!std.mem.eql(u8, tok.kind, "line_comment"));
        try std.testing.expect(!std.mem.eql(u8, tok.kind, "block_comment"));
    }
    // Should still have tokens from the function body
    try std.testing.expect(tokens.len > 0);
}

test "tokenizeTree: normalizes identifiers to V" {
    const parser = try tree_sitter.Parser.init();
    defer parser.deinit();
    try parser.setLanguage(.typescript);

    const source = "function myFunc(myParam: string): void { const myVar = myParam; }";
    const tree = try parser.parseString(source);
    defer tree.deinit();

    const root = tree.rootNode();
    const tokens = try tokenizeTree(std.testing.allocator, root, source);
    defer std.testing.allocator.free(tokens);

    // No token should retain the raw kind "identifier" — they must be normalized to "V"
    for (tokens) |tok| {
        try std.testing.expect(!std.mem.eql(u8, tok.kind, "identifier"));
    }
    // At least one token should be "V" (from the identifier nodes)
    var found_v = false;
    for (tokens) |tok| {
        if (std.mem.eql(u8, tok.kind, "V")) {
            found_v = true;
            break;
        }
    }
    try std.testing.expect(found_v);
}

test "tokenizeTree: skips type annotations" {
    const parser = try tree_sitter.Parser.init();
    defer parser.deinit();
    try parser.setLanguage(.typescript);

    // TypeScript version with type annotations
    const ts_source = "function f(x: number, y: string): boolean { return x > 0; }";
    // JavaScript equivalent without types
    const js_source = "function f(x, y) { return x > 0; }";

    const ts_tree = try parser.parseString(ts_source);
    defer ts_tree.deinit();
    const js_tree = try parser.parseString(js_source);
    defer js_tree.deinit();

    const ts_tokens = try tokenizeTree(std.testing.allocator, ts_tree.rootNode(), ts_source);
    defer std.testing.allocator.free(ts_tokens);
    const js_tokens = try tokenizeTree(std.testing.allocator, js_tree.rootNode(), js_source);
    defer std.testing.allocator.free(js_tokens);

    // TypeScript and JavaScript versions should produce the same token count
    // after stripping type annotations
    try std.testing.expectEqual(js_tokens.len, ts_tokens.len);
}

test "detectDuplication: finds clone groups in identical functions" {
    // Tokenize two copies of the same function under different file names
    const parser = try tree_sitter.Parser.init();
    defer parser.deinit();
    try parser.setLanguage(.typescript);

    const source_a =
        \\function processUserData(input: string): string {
        \\  const result = input.trim().toLowerCase();
        \\  if (result.length === 0) {
        \\    return "empty";
        \\  }
        \\  return result.split(",").join(";");
        \\}
    ;
    const source_b =
        \\function processItemData(input: string): string {
        \\  const result = input.trim().toLowerCase();
        \\  if (result.length === 0) {
        \\    return "empty";
        \\  }
        \\  return result.split(",").join(";");
        \\}
    ;

    const tree_a = try parser.parseString(source_a);
    defer tree_a.deinit();
    const tree_b = try parser.parseString(source_b);
    defer tree_b.deinit();

    const tokens_a = try tokenizeTree(std.testing.allocator, tree_a.rootNode(), source_a);
    defer std.testing.allocator.free(tokens_a);
    const tokens_b = try tokenizeTree(std.testing.allocator, tree_b.rootNode(), source_b);
    defer std.testing.allocator.free(tokens_b);

    const file_tokens = [_]FileTokens{
        .{ .path = "file_a.ts", .tokens = tokens_a },
        .{ .path = "file_b.ts", .tokens = tokens_b },
    };

    // Use a small min_window so the test functions (which may tokenize to ~20 tokens) are detected
    const config = DuplicationConfig{
        .min_window = 10,
        .file_warning_pct = 15.0,
        .file_error_pct = 25.0,
        .project_warning_pct = 5.0,
        .project_error_pct = 10.0,
    };
    const result = try detectDuplication(std.testing.allocator, &file_tokens, config);
    defer {
        for (result.clone_groups) |cg| {
            std.testing.allocator.free(cg.locations);
        }
        std.testing.allocator.free(result.clone_groups);
        std.testing.allocator.free(result.file_results);
    }

    // Should detect at least 1 clone group between these identical (modulo function name) functions
    try std.testing.expect(result.clone_groups.len >= 1);
}

test "detectDuplication: finds Type 2 clones with different identifiers" {
    // validateEmail and validatePhone are structurally identical but have different identifiers
    const parser = try tree_sitter.Parser.init();
    defer parser.deinit();
    try parser.setLanguage(.typescript);

    const source_email =
        \\function validateEmail(email: string): boolean {
        \\  const trimmed = email.trim();
        \\  if (trimmed.length === 0) {
        \\    return false;
        \\  }
        \\  return trimmed.includes("@");
        \\}
    ;
    const source_phone =
        \\function validatePhone(phone: string): boolean {
        \\  const cleaned = phone.trim();
        \\  if (cleaned.length === 0) {
        \\    return false;
        \\  }
        \\  return cleaned.includes("+");
        \\}
    ;

    const tree_e = try parser.parseString(source_email);
    defer tree_e.deinit();
    const tree_p = try parser.parseString(source_phone);
    defer tree_p.deinit();

    const tokens_e = try tokenizeTree(std.testing.allocator, tree_e.rootNode(), source_email);
    defer std.testing.allocator.free(tokens_e);
    const tokens_p = try tokenizeTree(std.testing.allocator, tree_p.rootNode(), source_phone);
    defer std.testing.allocator.free(tokens_p);

    const file_tokens = [_]FileTokens{
        .{ .path = "email.ts", .tokens = tokens_e },
        .{ .path = "phone.ts", .tokens = tokens_p },
    };

    const config = DuplicationConfig{
        .min_window = 8,
        .file_warning_pct = 15.0,
        .file_error_pct = 25.0,
        .project_warning_pct = 5.0,
        .project_error_pct = 10.0,
    };
    const result = try detectDuplication(std.testing.allocator, &file_tokens, config);
    defer {
        for (result.clone_groups) |cg| {
            std.testing.allocator.free(cg.locations);
        }
        std.testing.allocator.free(result.clone_groups);
        std.testing.allocator.free(result.file_results);
    }

    // Type 2 clones: structurally identical after identifier normalization
    try std.testing.expect(result.clone_groups.len >= 1);
}

test "detectDuplication: merges overlapping intervals correctly" {
    // When two clone windows overlap in the same file, duplication_pct must not exceed 100%
    const parser = try tree_sitter.Parser.init();
    defer parser.deinit();
    try parser.setLanguage(.typescript);

    const source =
        \\function processUserData(input: string): string {
        \\  const result = input.trim().toLowerCase();
        \\  if (result.length === 0) {
        \\    return "empty";
        \\  }
        \\  return result.split(",").join(";");
        \\}
        \\function processItemData(input: string): string {
        \\  const result = input.trim().toLowerCase();
        \\  if (result.length === 0) {
        \\    return "empty";
        \\  }
        \\  return result.split(",").join(";");
        \\}
    ;

    const tree = try parser.parseString(source);
    defer tree.deinit();

    const tokens = try tokenizeTree(std.testing.allocator, tree.rootNode(), source);
    defer std.testing.allocator.free(tokens);

    // Single file with duplicated functions inside it
    const file_tokens = [_]FileTokens{
        .{ .path = "combined.ts", .tokens = tokens },
    };

    const config = DuplicationConfig{
        .min_window = 8,
        .file_warning_pct = 15.0,
        .file_error_pct = 25.0,
        .project_warning_pct = 5.0,
        .project_error_pct = 10.0,
    };
    const result = try detectDuplication(std.testing.allocator, &file_tokens, config);
    defer {
        for (result.clone_groups) |cg| {
            std.testing.allocator.free(cg.locations);
        }
        std.testing.allocator.free(result.clone_groups);
        std.testing.allocator.free(result.file_results);
    }

    // duplication_pct must never exceed 100%
    for (result.file_results) |fr| {
        try std.testing.expect(fr.duplication_pct <= 100.0);
    }
    try std.testing.expect(result.project_duplication_pct <= 100.0);
}

File: src/metrics/duplication.rs
use rustc_hash::FxHashMap;

use crate::metrics::halstead::is_type_only_node;
use crate::types::{CloneGroup, CloneInstance, DuplicationConfig, DuplicationResult, Token};

/// Rolling hash base constant matching Zig implementation.
const HASH_BASE: u64 = 37;

/// Maximum bucket size for hash buckets. Buckets larger than this are likely
/// common patterns (boilerplate, short keywords) and are discarded to prevent
/// O(N^2) verification.
const MAX_BUCKET_SIZE: usize = 1000;

/// Tokenize a tree-sitter AST into a normalized token sequence.
///
/// Strips comments, whitespace, structural punctuation, and TypeScript type annotations.
/// Normalizes identifier nodes to sentinel "V" for Type 2 clone detection.
/// Each token stores a precomputed hash for efficient rolling hash computation.
pub fn tokenize_tree(root: tree_sitter::Node, source: &[u8]) -> Vec<Token> {
    let mut tokens = Vec::new();
    tokenize_node(root, source, &mut tokens);
    tokens
}

/// Recursively collect normalized tokens from an AST node.
#[allow(clippy::only_used_in_recursion)]
fn tokenize_node(node: tree_sitter::Node, source: &[u8], tokens: &mut Vec<Token>) {
    let kind = node.kind();

    // Skip entire TypeScript type annotation subtrees
    if is_type_only_node(kind) {
        return;
    }

    if node.child_count() == 0 {
        // Leaf node
        if is_skipped_kind(kind) {
            return;
        }

        let normalized = normalize_kind(kind);
        tokens.push(Token {
            kind: normalized,
            kind_hash: token_hash(normalized),
            start_byte: node.start_byte(),
            end_byte: node.end_byte(),
        });
        return;
    }

    // Non-leaf: recurse into all children
    for i in 0..node.child_count() as u32 {
        if let Some(child) = node.child(i) {
            tokenize_node(child, source, tokens);
        }
    }
}

/// Returns true for token kinds that should be skipped during tokenization.
fn is_skipped_kind(kind: &str) -> bool {
    matches!(
        kind,
        "comment" | "line_comment" | "block_comment" | ";" | "," | "hash_bang_line"
    )
}

/// Returns the normalized kind for a token.
/// Identifiers (all variants) are normalized to sentinel "V" for Type 2 clone detection.
/// Returns `&'static str` since all inputs come from tree-sitter (static grammar strings).
fn normalize_kind(kind: &'static str) -> &'static str {
    match kind {
        "identifier"
        | "property_identifier"
        | "shorthand_property_identifier"
        | "shorthand_property_identifier_pattern" => "V",
        other => other,
    }
}

/// Compute a hash value for a token kind string using Rabin-Karp polynomial hashing.
fn token_hash(kind: &str) -> u64 {
    let mut h: u64 = 0;
    for &c in kind.as_bytes() {
        h = h.wrapping_mul(HASH_BASE).wrapping_add(c as u64);
    }
    h
}

/// Rolling hash state for a sliding window over a token sequence.
struct RollingHasher {
    hash: u64,
    /// B^(window_size - 1) for removing the leftmost token.
    base_pow: u64,
}

impl RollingHasher {
    /// Initialize the hasher over the first `window` tokens.
    fn new(tokens: &[Token], window: usize) -> Self {
        let mut h: u64 = 0;
        let mut bpow: u64 = 1;
        for (i, token) in tokens.iter().enumerate().take(window) {
            h = h.wrapping_mul(HASH_BASE).wrapping_add(token.kind_hash);
            if i < window - 1 {
                bpow = bpow.wrapping_mul(HASH_BASE);
            }
        }
        RollingHasher {
            hash: h,
            base_pow: bpow,
        }
    }

    /// Slide the window: remove left token, add right token.
    fn roll(&mut self, remove: &Token, add: &Token) {
        self.hash = self
            .hash
            .wrapping_sub(remove.kind_hash.wrapping_mul(self.base_pow))
            .wrapping_mul(HASH_BASE)
            .wrapping_add(add.kind_hash);
    }
}

/// Detect code clones across multiple files using Rabin-Karp rolling hash.
pub fn detect_duplication(
    file_tokens: &[&[Token]],
    config: &DuplicationConfig,
) -> DuplicationResult {
    let window = config.min_tokens as usize;

    if file_tokens.is_empty() || window == 0 {
        return DuplicationResult {
            clone_groups: Vec::new(),
            total_tokens: 0,
            cloned_tokens: 0,
            duplication_percentage: 0.0,
        };
    }

    // Build hash index: hash -> Vec<(file_index, start_token_position)>
    let mut index: FxHashMap<u64, Vec<(usize, usize)>> = FxHashMap::default();

    for (file_idx, tokens) in file_tokens.iter().enumerate() {
        if tokens.len() < window {
            continue;
        }

        let mut hasher = RollingHasher::new(tokens, window);
        let mut start = 0;

        while start + window <= tokens.len() {
            let bucket = index.entry(hasher.hash).or_default();
            bucket.push((file_idx, start));

            if start + window < tokens.len() {
                hasher.roll(&tokens[start], &tokens[start + window]);
            }
            start += 1;
        }
    }

    // Form clone groups from verified hash collisions
    let mut clone_groups: Vec<CloneGroup> = Vec::new();

    // Per-file interval lists for counting cloned tokens
    let num_files = file_tokens.len();
    let mut intervals_per_file: Vec<Vec<(usize, usize)>> = vec![Vec::new(); num_files];

    for bucket in index.values() {
        if bucket.len() < 2 || bucket.len() > MAX_BUCKET_SIZE {
            continue;
        }

        let mut instances: Vec<CloneInstance> = Vec::new();
        let mut added: Vec<bool> = vec![false; bucket.len()];

        for i in 0..bucket.len() {
            for j in (i + 1)..bucket.len() {
                let (fi_a, st_a) = bucket[i];
                let (fi_b, st_b) = bucket[j];

                // Skip identical position in same file
                if fi_a == fi_b && st_a == st_b {
                    continue;
                }

                // Verify token-by-token match
                if tokens_match(file_tokens[fi_a], st_a, file_tokens[fi_b], st_b, window) {
                    if !added[i] {
                        added[i] = true;
                        instances.push(make_instance(fi_a, st_a, window, file_tokens[fi_a]));
                        intervals_per_file[fi_a].push((st_a, st_a + window));
                    }
                    if !added[j] {
                        added[j] = true;
                        instances.push(make_instance(fi_b, st_b, window, file_tokens[fi_b]));
                        intervals_per_file[fi_b].push((st_b, st_b + window));
                    }
                }
            }
        }

        if instances.len() >= 2 {
            clone_groups.push(CloneGroup {
                instances,
                token_count: window as u32,
            });
        }
    }

    // Count total tokens and cloned tokens (with interval merging)
    let mut total_tokens: usize = 0;
    let mut cloned_tokens: usize = 0;

    for (file_idx, tokens) in file_tokens.iter().enumerate() {
        total_tokens += tokens.len();
        cloned_tokens += count_merged_intervals(&mut intervals_per_file[file_idx]);
    }

    let duplication_percentage = if total_tokens == 0 {
        0.0
    } else {
        cloned_tokens as f64 / total_tokens as f64 * 100.0
    };

    DuplicationResult {
        clone_groups,
        total_tokens,
        cloned_tokens,
        duplication_percentage,
    }
}

/// Create a CloneInstance from file tokens.
fn make_instance(
    file_index: usize,
    start_token: usize,
    window: usize,
    _tokens: &[Token],
) -> CloneInstance {
    CloneInstance {
        file_index,
        start_token,
        end_token: start_token + window,
        start_line: 0,
        end_line: 0,
    }
}

/// Verify that two token windows have identical normalized token sequences.
/// Uses pointer comparison on static strings for speed, with string fallback.
fn tokens_match(
    a_tokens: &[Token],
    a_start: usize,
    b_tokens: &[Token],
    b_start: usize,
    window: usize,
) -> bool {
    for i in 0..window {
        let a = a_tokens[a_start + i].kind;
        let b = b_tokens[b_start + i].kind;
        // Fast path: pointer equality for static strings from same grammar
        if !std::ptr::eq(a.as_ptr(), b.as_ptr()) && a != b {
            return false;
        }
    }
    true
}

/// Merge overlapping intervals and return total non-overlapping token count.
fn count_merged_intervals(intervals: &mut [(usize, usize)]) -> usize {
    if intervals.is_empty() {
        return 0;
    }

    intervals.sort_by_key(|&(start, _)| start);

    let mut total: usize = 0;
    let mut cur_start = intervals[0].0;
    let mut cur_end = intervals[0].1;

    for &(start, end) in intervals.iter().skip(1) {
        if start <= cur_end {
            // Overlapping or adjacent
            if end > cur_end {
                cur_end = end;
            }
        } else {
            // Gap
            total += cur_end - cur_start;
            cur_start = start;
            cur_end = end;
        }
    }
    total += cur_end - cur_start;

    total
}

// TESTS

#[cfg(test)]
mod tests {
    use super::*;
    use std::path::Path;

    fn parse_to_tokens(source: &str) -> Vec<Token> {
        let language: tree_sitter::Language = tree_sitter_typescript::LANGUAGE_TYPESCRIPT.into();
        let mut parser = tree_sitter::Parser::new();
        parser.set_language(&language).unwrap();
        let tree = parser.parse(source.as_bytes(), None).unwrap();
        tokenize_tree(tree.root_node(), source.as_bytes())
    }

    #[test]
    fn tokenize_produces_tokens() {
        let source = "function add(a: number, b: number): number { return a + b; }";
        let tokens = parse_to_tokens(source);
        assert!(
            !tokens.is_empty(),
            "should produce tokens from a real function"
        );
    }

    #[test]
    fn tokenize_skips_comments() {
        let source = r#"
// This is a comment
function greet(name: string): string {
  /* block comment */
  return name;
}
"#;
        let tokens = parse_to_tokens(source);
        for tok in &tokens {
            assert_ne!(tok.kind, "comment");
            assert_ne!(tok.kind, "line_comment");
            assert_ne!(tok.kind, "block_comment");
        }
        assert!(!tokens.is_empty());
    }

    #[test]
    fn tokenize_normalizes_identifiers_to_v() {
        let source = "function myFunc(myParam: string): void { const myVar = myParam; }";
        let tokens = parse_to_tokens(source);
        // No raw "identifier" kind should remain
        for tok in &tokens {
            assert_ne!(
                tok.kind, "identifier",
                "identifier should be normalized to V"
            );
        }
        // At least one "V" token
        assert!(tokens.iter().any(|t| t.kind == "V"), "should have V tokens");
    }

    #[test]
    fn tokenize_skips_type_annotations() {
        // TS with types
        let ts_source = "function f(x: number, y: string): boolean { return x > 0; }";
        // JS without types
        let js_source = "function f(x, y) { return x > 0; }";

        let ts_tokens = parse_to_tokens(ts_source);
        let js_tokens = parse_to_tokens(js_source);

        assert_eq!(
            ts_tokens.len(),
            js_tokens.len(),
            "TS and JS should produce same token count after type stripping"
        );
    }

    #[test]
    fn token_hash_deterministic() {
        let h1 = token_hash("function");
        let h2 = token_hash("function");
        assert_eq!(h1, h2, "same input should produce same hash");

        let h3 = token_hash("return");
        assert_ne!(h1, h3, "different inputs should produce different hashes");
    }

    #[test]
    fn is_skipped_kind_filters_correctly() {
        assert!(is_skipped_kind("comment"));
        assert!(is_skipped_kind("line_comment"));
        assert!(is_skipped_kind("block_comment"));
        assert!(is_skipped_kind(";"));
        assert!(is_skipped_kind(","));
        assert!(is_skipped_kind("hash_bang_line"));
        assert!(!is_skipped_kind("identifier"));
        assert!(!is_skipped_kind("function"));
    }

    #[test]
    fn detect_duplication_finds_clones_in_identical_functions() {
        let source_a = r#"function processUserData(input: string): string {
  const result = input.trim().toLowerCase();
  if (result.length === 0) {
    return "empty";
  }
  return result.split(",").join(";");
}"#;
        let source_b = r#"function processItemData(input: string): string {
  const result = input.trim().toLowerCase();
  if (result.length === 0) {
    return "empty";
  }
  return result.split(",").join(";");
}"#;

        let tokens_a = parse_to_tokens(source_a);
        let tokens_b = parse_to_tokens(source_b);

        let config = DuplicationConfig {
            min_tokens: 10,
            enabled: true,
        };
        let result = detect_duplication(&[tokens_a.as_slice(), tokens_b.as_slice()], &config);
        assert!(
            !result.clone_groups.is_empty(),
            "should detect clones between identical functions"
        );
    }

    #[test]
    fn detect_duplication_type2_different_identifiers() {
        let source_email = r#"function validateEmail(email: string): boolean {
  const trimmed = email.trim();
  if (trimmed.length === 0) {
    return false;
  }
  return trimmed.includes("@");
}"#;
        let source_phone = r#"function validatePhone(phone: string): boolean {
  const cleaned = phone.trim();
  if (cleaned.length === 0) {
    return false;
  }
  return cleaned.includes("+");
}"#;

        let tokens_e = parse_to_tokens(source_email);
        let tokens_p = parse_to_tokens(source_phone);

        let config = DuplicationConfig {
            min_tokens: 8,
            enabled: true,
        };
        let result = detect_duplication(&[tokens_e.as_slice(), tokens_p.as_slice()], &config);
        assert!(
            !result.clone_groups.is_empty(),
            "should detect Type 2 clones with different identifiers"
        );
    }

    #[test]
    fn detect_duplication_interval_merging() {
        let source = r#"function processUserData(input: string): string {
  const result = input.trim().toLowerCase();
  if (result.length === 0) {
    return "empty";
  }
  return result.split(",").join(";");
}
function processItemData(input: string): string {
  const result = input.trim().toLowerCase();
  if (result.length === 0) {
    return "empty";
  }
  return result.split(",").join(";");
}"#;

        let tokens = parse_to_tokens(source);

        let config = DuplicationConfig {
            min_tokens: 8,
            enabled: true,
        };
        let result = detect_duplication(&[tokens.as_slice()], &config);
        assert!(
            result.duplication_percentage <= 100.0,
            "duplication_pct must not exceed 100%, got {}",
            result.duplication_percentage
        );
    }

    #[test]
    fn tokenize_duplication_fixture() {
        let fixture_path = Path::new(env!("CARGO_MANIFEST_DIR"))
            .join("tests/fixtures/typescript/duplication_cases.ts");
        let source = std::fs::read_to_string(&fixture_path).unwrap();
        let tokens = parse_to_tokens(&source);

        assert!(!tokens.is_empty(), "fixture should produce tokens");
        // All identifiers should be normalized to "V"
        for tok in &tokens {
            assert_ne!(tok.kind, "identifier", "identifiers should be normalized");
            assert_ne!(
                tok.kind, "property_identifier",
                "property_identifiers should be normalized"
            );
        }
        // Should have at least some "V" tokens
        assert!(tokens.iter().any(|t| t.kind == "V"), "should have V tokens");
    }

    #[test]
    fn count_merged_intervals_basic() {
        let mut intervals = vec![(0, 5), (3, 8), (10, 15)];
        let count = count_merged_intervals(&mut intervals);
        // (0,5)+(3,8) merge to (0,8)=8, plus (10,15)=5, total=13
        assert_eq!(count, 13);
    }

    #[test]
    fn count_merged_intervals_no_overlap() {
        let mut intervals = vec![(0, 3), (5, 8), (10, 12)];
        let count = count_merged_intervals(&mut intervals);
        assert_eq!(count, 3 + 3 + 2);
    }

    #[test]
    fn count_merged_intervals_empty() {
        let mut intervals: Vec<(usize, usize)> = vec![];
        let count = count_merged_intervals(&mut intervals);
        assert_eq!(count, 0);
    }
}
